kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -0,0 +1,254 @@
+            StructType readSchema) {
+        this.tableClient = tableClient;
+        this.filesIter = files.iterator();
+        this.readSchema = readSchema;
+        this.actionsIter = Optional.empty();
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (closed) {
+            throw new IllegalStateException("Can't call `hasNext` on a closed iterator.");
+        }
+
+        tryEnsureNextActionsIterIsReady();
+
+        // By definition of tryEnsureNextActionsIterIsReady, we know that if actionsIter
+        // is non-empty then it has a next element
+
+        return actionsIter.isPresent();
+    }
+
+    /**
+     * @return a tuple of (FileDataReadResult, isFromCheckpoint), where FileDataReadResult conforms
+     *         to the instance {@link #readSchema}.
+     */
+    @Override
+    public Tuple2<FileDataReadResult, Boolean> next() {
+        if (closed) {
+            throw new IllegalStateException("Can't call `next` on a closed iterator.");
+        }
+
+        if (!hasNext()) throw new NoSuchElementException("No next element");
+
+        return actionsIter.get().next();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (!closed && actionsIter.isPresent()) {
+            actionsIter.get().close();
+            actionsIter = Optional.empty();
+            closed = true;
+        }
+    }
+
+    /**
+     * If the current `actionsIter` has no more elements, this function finds the next
+     * non-empty file in `filesIter` and uses it to set `actionsIter`.
+     */
+    private void tryEnsureNextActionsIterIsReady() {
+        if (actionsIter.isPresent()) {
+            // This iterator already has a next element, so we can exit early;
+            if (actionsIter.get().hasNext()) {
+                return;
+            }
+
+            // Clean up resources
+            Utils.closeCloseables(actionsIter.get());
+
+            // Set this to empty since we don't know if there's a next file yet
+            actionsIter = Optional.empty();
+        }
+
+        // Search for the next non-empty file and use that iter
+        while (filesIter.hasNext()) {
+            actionsIter = Optional.of(getNextActionsIter());
+
+            if (actionsIter.get().hasNext()) {
+                // It is ready, we are done
+                return;
+            }
+
+            // It was an empty file. Clean up resources
+            Utils.closeCloseables(actionsIter.get());
+
+            // Set this to empty since we don't know if there's a next file yet
+            actionsIter = Optional.empty();
+        }
+    }
+
+    /**
+     * Get the next file from `filesIter` (.json or .checkpoint.parquet), contextualize it
+     * (allow the connector to split it), and then read it + inject the `isFromCheckpoint`
+     * information.
+     *
+     * Requires that `filesIter.hasNext` is true.
+     */
+    private CloseableIterator<Tuple2<FileDataReadResult, Boolean>> getNextActionsIter() {
```

Comment:
We discussed [this](https://github.com/delta-io/delta/pull/1939#discussion_r1284607044). If this pattern iterator flattening pattern pops up again, we will make the it an utility. something IteratorUtils.flatten.

Reply:
Code reuse was only a secondary motivation. This sort of "iterator that does all the things" code is notoriously difficult to get right because there are so many corner cases to reason about. Splitting up the logic changes O(n*m) complexity to O(n+m).